pkg/params/lister.go

```diff
@@ -0,0 +1,280 @@
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package params
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"sort"
+
+	"github.com/google/go-jsonnet/ast"
+	"github.com/ksonnet/ksonnet-lib/ksonnet-gen/astext"
+	"github.com/ksonnet/ksonnet-lib/ksonnet-gen/printer"
+	"github.com/ksonnet/ksonnet/pkg/app"
+	"github.com/ksonnet/ksonnet/pkg/util/jsonnet"
+	"github.com/pkg/errors"
+)
+
+// Entry is parameter entry. It uses to hold output when listing
+// parameters.
+type Entry struct {
+	// ComponentName is the component that owns this entry.
+	ComponentName string
+	// ParamName is the name of the parameter.
+	ParamName string
+	// Value is the value of the parameter.
+	Value string
+}
+
+// Lister lists parameters.
+type Lister struct {
+	// Destination is the cluster details.
+	Destination app.EnvironmentDestinationSpec
+	// AppRoot is the root path for the application.
+	AppRoot string
+
+	createEntry func(id string, object *astext.Object) ([]Entry, error)
+}
+
+// NewLister creates an instance of Lister.
+func NewLister(appRoot string, destination app.EnvironmentDestinationSpec) *Lister {
+	ec := newEntryCreator()
+
+	return &Lister{
+		AppRoot:     appRoot,
+		Destination: destination,
+
+		createEntry: ec.Create,
+	}
+}
+
+// List lists parameters in a slice of Entry.
+func (l *Lister) List(r io.Reader, componentName string) ([]Entry, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading parameters source")
+	}
+
+	source := string(data)
+
+	object, err := l.buildObject(source)
+	if err != nil {
+		return nil, errors.Wrap(err, "building params object")
+	}
+
+	co, err := l.componentsObject(object)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding parameter components")
+	}
+
+	var entries []Entry
+
+	for i := range co.Fields {
+		f := co.Fields[i]
+
+		id, err := jsonnet.FieldID(f)
+		if err != nil {
+			return nil, errors.Wrap(err, "retrieving field from object")
+		}
+
+		if componentName != "" && componentName != id {
+			// if filter is based on a component name, continue if this component is
+			// not in the list
+			continue
+		}
+
+		paramObject, ok := f.Expr2.(*astext.Object)
+		if !ok {
+			return nil, errors.Errorf("component %q value is not an object", id)
+		}
+
+		paramEntries, err := l.createEntry(id, paramObject)
+		if err != nil {
+			return nil, err
+		}
+
+		entries = append(entries, paramEntries...)
+	}
+
+	sort.Slice(entries, func(i, j int) bool {
+		if entries[i].ComponentName < entries[j].ComponentName {
+			return true
+		}
+
+		return entries[i].ParamName < entries[j].ParamName
+	})
+
+	return entries, nil
+}
+
+// componentsObject finds the components object in side of a Jsonnet object in params format.
+func (l *Lister) componentsObject(object *astext.Object) (*astext.Object, error) {
+	for i := range object.Fields {
```

Comment:
This is not significant, but I was wondering why not use the two valued variant of `range`?

Reply:
Because I want to copy the field out rather than reusing the variable.

Reply:
The variable is reused, but isn't the map value being copied over it each iteration?
I had trouble finding the justification in the spec, but that was my understanding.

https://play.golang.org/p/Cp5Wb6MliLe

It's possible I'm misunderstanding.